### ISSUE TITLE
Cloning an trigger and enabling/disabling the clone many times in the same frame causes Ammo crash

### ISSUE THREAD

**Issue by yaustar** [playcanvas/engine#2572]
Repo: https://playcanvas.com/editor/scene/1037430
Press 1, then 2, then 1 again. It should crash then.

I can't explain this one, there's some combination of disabling on initialisation and enabling/disabling on clone causes some odd behaviour.

Moving 
```
// initialize code called once per entity
DisableMeOnInit.prototype.initialize = function() {
    this.entity.enabled = false;
};
``` 
To postInitialize fixes the issue:
```
// initialize code called once per entity
DisableMeOnInit.prototype.postInitialize = function() {
    this.entity.enabled = false;
};
```

**Comment by LeXXik** [comment]
I think there is something to do with the order of enabling/disabling the trigger component. I noticed after pressing 1, which should add a clone, the `enable` method ran twice, adding 2 triggers. As the error happened over and over, the `_triggers` array kept growing with new triggers:
![image](https://user-images.githubusercontent.com/5677782/99904094-7f45cd00-2cd1-11eb-832c-bd0b1a71c04d.png)

**Comment by kungfooman** [comment]
Took me some time to find actual project so people can fork and download to test locally: https://playcanvas.com/project/739702/overview/trigger-crash

**Comment by yaustar** [comment]
@kungfooman Tip, if you only have a scene link like the first post, you can go to the project homepage view this button 
<img width="720" alt="image" src="https://github.com/user-attachments/assets/999bc16c-366b-40db-8a7e-21c5be030b92">

**Comment by kungfooman** [comment]
@yaustar Funny, thank you! Sometimes I don't get the most obvious things :zany_face:

**Comment by kungfooman** [comment]
I don't know why it happens, but just pointing out the obvious so far:

If you look at the wasm position where it crashes and when it crashes:

![image](https://github.com/user-attachments/assets/3c1cc6fd-700a-4ce3-8f15-2e572daac8e6)

Another bad call:

![image](https://github.com/user-attachments/assets/ada7b86b-f0da-4619-b28a-1b3a36f99c4d)

Good calls look like:

![image](https://github.com/user-attachments/assets/04a8b02a-05db-4370-b396-8b78e5154d92)

![image](https://github.com/user-attachments/assets/d72911a7-ce35-41ac-baf9-7325f7513417)

So basically the bad calls contain a `NULL` pointer and a pointer that looks like garbage (any negative/positive super large number).

Probably some kind of memory corruption or something got freed to early?

**Comment by LeXXik** [comment]
Briefly looked at it. The crash in Ammo happens when it tries to update AABB of a single trigger and fails to find it. The issue is somewhere in combination of disabling a cloned entity and the fact that we don't destroy trigger when disable an entity (we only remove it from the dynamics world). If I destroy the trigger on entity disable (like we do with normal rigid bodies), then the error goes away.

**Comment by LeXXik** [comment]
I tried to replicate with just creating a body and then randomly adding/removing/updating it. Basically stuff that happens, when we enable/disable entity and update the systems. The crash doesn't happen :/

```js
const app = this.app;

const transform = new Ammo.btTransform();
const vec = new Ammo.btVector3(0.5, 0.5, 0.5);
const quat = new Ammo.btQuaternion(0, 0, 0, 1);
transform.setOrigin(vec);
transform.setRotation(quat);

const shape = new Ammo.btBoxShape(vec);
const body = this.app.systems.rigidbody.createBody(1, shape, transform);

vec.setValue(0, 0, 0);
body.setRestitution(0);
body.setFriction(0);
body.setDamping(0, 0);
body.setLinearFactor(vec);
body.setAngularFactor(vec);
body.setCollisionFlags(body.getCollisionFlags() | pc.BODYFLAG_NORESPONSE_OBJECT);


remove();
add();
add();
remove();
updateTransform();
add();
updateTransform();
remove();
remove();
updateTransform();


function add() {
    app.systems.rigidbody.addBody(body, pc.BODYGROUP_TRIGGER, pc.BODYMASK_NOT_STATIC ^ pc.BODYGROUP_TRIGGER);
    body.forceActivationState(pc.BODYSTATE_ACTIVE_TAG);

    console.log('added');
}

function remove() {
    app.systems.rigidbody.removeBody(body);
    body.forceActivationState(pc.BODYSTATE_DISABLE_SIMULATION);

    console.log('removed');
}

function updateTransform() {
    vec.setValue(Math.random(), Math.random(), Math.random());
    quat.setValue(Math.random(), Math.random(), Math.random(), 1);
    quat.normalize();
    transform.setOrigin(vec);
    transform.setRotation(quat);

    body.setWorldTransform(transform);
    body.activate();

    console.log('updated');
}

Ammo.destroy(transform);
Ammo.destroy(vec);
Ammo.destroy(quat);
```

**Comment by LeXXik** [comment]
From the snippet above, it looks like the issue in the example happens here:

![image](https://github.com/user-attachments/assets/6f0bcca6-2023-4f6c-9264-46e395aa3698)

Every time we enable a trigger, we add it to the dynamics world, regardless if we already added it there before or not. This results in the body being added twice.

https://github.com/playcanvas/engine/blob/a1e66ee179ccaf806709910be4f6ffb784afece7/src/framework/components/collision/trigger.js#L118-L124

**Comment by LeXXik** [comment]
So, the `disable()` is not called when `entity.enabled = false` because of that `disable-me` script, which disables parent. Steps:

```
1. clone - create trigger

2. entity.enabled = true:
2.1. enable disable-me script - remove trigger from world, set entity to disabled
2.2. add trigger to world

3. entity.enabled = false:
3.1. trigger entity already disabled, do nothing

4. entity.enabled = true:
4.1 add trigger to world
```

Which results in trigger being added twice.

**Comment by LeXXik** [comment]
Possible solution: check if trigger already exists in the array, before adding it:

```js
const exists = systems.rigidbody._triggers.find(trigger => trigger === this);

if (!exists) {
    systems.rigidbody.addBody(body, BODYGROUP_TRIGGER, BODYMASK_NOT_STATIC ^ BODYGROUP_TRIGGER);
    systems.rigidbody._triggers.push(this);
}
```
